File: Python/백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
import sys
N = int(sys.stdin.readline())
inp = list(map(int,sys.stdin.readline().split()))
cnt = [0 for _ in range(N)]
cnt[0] = 1
for i in range(1, N):
	M = 0
	for j in range(0, i):
		if inp[i] > inp[j]:
			M = max(M, cnt[j])
	cnt[i] = M + 1
print(max(cnt))

File: Java/백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.io.*;
import java.util.*;

class Solution {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public void solution() throws IOException{
        int N = Integer.parseInt(br.readLine());
        int inps[] = new int[N];
        String inp[] = br.readLine().split(" ");
        for (int i = 0; i < N; i++) inps[i] = Integer.parseInt(inp[i]);

        int ans[] = new int[N];
        Arrays.fill(ans, 1);

        for (int i = 0; i < N - 1; i++) {
            for (int j = i + 1; j < N; j++) {
                if (inps[j] > inps[i]) ans[j] = Integer.max(ans[j], ans[i] + 1);
            }
        }
        int maxInt = Integer.MIN_VALUE;
        for (int i = 0; i < N; i++) maxInt = Integer.max(maxInt, ans[i]);
        bw.write(maxInt + "\n");
        bw.flush();
        bw.close();
    }
}

public class Main {
    public static void main(String argv[]) throws IOException{
        Solution sol = new Solution();
        sol.solution();
    }
}